ambry-frontend/src/test/java/com.github.ambry.frontend/FrontendIntegrationTest.java

```diff
@@ -357,24 +366,46 @@ private String postBlobAndVerify(HttpHeaders headers, ByteBuffer content)
   /**
    * Gets the blob with blob ID {@code blobId} and verifies that the headers and content match with what is expected.
    * @param blobId the blob ID of the blob to GET.
+   * @param range the {@link ByteRange} for the request.
    * @param expectedHeaders the expected headers in the response.
    * @param expectedContent the expected content of the blob.
    * @throws ExecutionException
    * @throws InterruptedException
    */
-  private void getBlobAndVerify(String blobId, HttpHeaders expectedHeaders, ByteBuffer expectedContent)
+  private void getBlobAndVerify(String blobId, ByteRange range, HttpHeaders expectedHeaders, ByteBuffer expectedContent)
       throws ExecutionException, InterruptedException {
-    FullHttpRequest httpRequest = buildRequest(HttpMethod.GET, blobId, null, null);
+    HttpHeaders headers = null;
+    if (range != null) {
+      headers = new DefaultHttpHeaders().add(RestUtils.Headers.RANGE, RestTestUtils.getRangeHeaderString(range));
+    }
+    FullHttpRequest httpRequest = buildRequest(HttpMethod.GET, blobId, headers, null);
     Queue<HttpObject> responseParts = nettyClient.sendRequest(httpRequest, null, null).get();
     HttpResponse response = (HttpResponse) responseParts.poll();
-    assertEquals("Unexpected response status", HttpResponseStatus.OK, response.getStatus());
+    assertEquals("Unexpected response status",
+        range == null ? HttpResponseStatus.OK : HttpResponseStatus.PARTIAL_CONTENT, response.getStatus());
     checkCommonGetHeadHeaders(response.headers());
     assertEquals("Content-Type does not match", expectedHeaders.get(RestUtils.Headers.AMBRY_CONTENT_TYPE),
         response.headers().get(HttpHeaders.Names.CONTENT_TYPE));
     assertEquals(RestUtils.Headers.BLOB_SIZE + " does not match", expectedHeaders.get(RestUtils.Headers.BLOB_SIZE),
         response.headers().get(RestUtils.Headers.BLOB_SIZE));
-    ByteBuffer responseContent = getContent(response, responseParts);
-    assertArrayEquals("GET content does not match original content", expectedContent.array(), responseContent.array());
+    assertEquals("Accept-Ranges not set correctly", "bytes", response.headers().get(RestUtils.Headers.ACCEPT_RANGES));
+    byte[] expectedContentArray = expectedContent.array();
+    if (range != null) {
+      long blobSize = Long.parseLong(expectedHeaders.get(RestUtils.Headers.BLOB_SIZE));
+      ByteRange resolvedRange = range.toResolvedByteRange(blobSize);
+      assertEquals("Content-Range header not set correctly", RestUtils.buildContentRangeHeader(resolvedRange, blobSize),
+          response.headers().get(RestUtils.Headers.CONTENT_RANGE));
+      expectedContentArray = Arrays.copyOfRange(expectedContentArray, (int) resolvedRange.getStartOffset(),
+          (int) resolvedRange.getEndOffset() + 1);
+    } else {
+      assertNull("Content-Range header should not be set", response.headers().get(RestUtils.Headers.CONTENT_RANGE));
+    }
+    if (HttpHeaders.isContentLengthSet(response)) {
```

Comment:
I think you should first determine whether content-length should be set in the test itself and check for the content length (if that is possible). You can't rely on the server having made the right decision.